staging/src/k8s.io/client-go/util/workqueue/delaying_queue_test.go

```diff
@@ -135,17 +408,15 @@ func TestAddTwoFireEarly(t *testing.T) {
 
 	q.AddAfter(first, 1*time.Second)
```

Comment:
Can you please try to [stress run](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/flaky-tests.md#deflaking-unit-tests) the queue tests?

I have found the following flake in this test:

```
/tmp/go-stress-20230704T005647-845419652
--- FAIL: TestAddTwoFireEarly (2.01s)
    delaying_queue_test.go:441: unexpected err: context deadline exceeded
FAIL
```

And also another flake (run on the whole workqueue package):

```
SIGABRT: abort
PC=0x4a99c1 m=0 sigcode=0

goroutine 0 [idle]:
runtime.futex()
	/usr/lib/go/src/runtime/sys_linux_amd64.s:554 +0x21 fp=0x7ffeeaf2e628 sp=0x7ffeeaf2e620 pc=0x4a99c1
runtime.futexsleep(0x10000000000?, 0xeaf2e6a8?, 0x47e96c?)
	/usr/lib/go/src/runtime/os_linux.go:69 +0x36 fp=0x7ffeeaf2e678 sp=0x7ffeeaf2e628 pc=0x46b5d6
runtime.notesleep(0xb2af08)
	/usr/lib/go/src/runtime/lock_futex.go:160 +0x87 fp=0x7ffeeaf2e6b0 sp=0x7ffeeaf2e678 pc=0x444487
runtime.mPark(...)
	/usr/lib/go/src/runtime/proc.go:1530
runtime.stopm()
	/usr/lib/go/src/runtime/proc.go:2334 +0x8c fp=0x7ffeeaf2e6e0 sp=0x7ffeeaf2e6b0 pc=0x47614c
runtime.findRunnable()
	/usr/lib/go/src/runtime/proc.go:3007 +0xabc fp=0x7ffeeaf2e7e8 sp=0x7ffeeaf2e6e0 pc=0x47799c
runtime.schedule()
	/usr/lib/go/src/runtime/proc.go:3360 +0xb1 fp=0x7ffeeaf2e820 sp=0x7ffeeaf2e7e8 pc=0x4787d1
runtime.park_m(0xc000111d40?)
	/usr/lib/go/src/runtime/proc.go:3511 +0x12d fp=0x7ffeeaf2e850 sp=0x7ffeeaf2e820 pc=0x478ced
runtime.mcall()
	/usr/lib/go/src/runtime/asm_amd64.s:452 +0x43 fp=0x7ffeeaf2e860 sp=0x7ffeeaf2e850 pc=0x4a5163

goroutine 1 [chan receive, 9 minutes]:
runtime.gopark(0xc0001a3680?, 0x445bb5?, 0x50?, 0x18?, 0x0?)
	/usr/lib/go/src/runtime/proc.go:381 +0xd6 fp=0xc0001a3600 sp=0xc0001a35e0 pc=0x472236
runtime.chanrecv(0xc00022e070, 0xc0001a36fe, 0x1)
	/usr/lib/go/src/runtime/chan.go:583 +0x42e fp=0xc0001a3690 sp=0xc0001a3600 pc=0x43e88e
runtime.chanrecv1(0xb29ec0?, 0x82d6e0?)
	/usr/lib/go/src/runtime/chan.go:442 +0x18 fp=0xc0001a36b8 sp=0xc0001a3690 pc=0x43e3f8
testing.(*T).Run(0xc00008b040, {0x89a647, 0x23}, 0x8beb18)
	/usr/lib/go/src/testing/testing.go:1630 +0x82e fp=0xc0001a37f8 sp=0xc0001a36b8 pc=0x5bacae
testing.runTests.func1(0x0?)
	/usr/lib/go/src/testing/testing.go:2036 +0x8e fp=0xc0001a3850 sp=0xc0001a37f8 pc=0x5be80e
testing.tRunner(0xc00008b040, 0xc0001a3a98)
	/usr/lib/go/src/testing/testing.go:1576 +0x217 fp=0xc0001a3958 sp=0xc0001a3850 pc=0x5b9437
testing.runTests(0xc0000c6be0?, {0xb24b40, 0x1f, 0x1f}, {0xc0000ae43c?, 0xc0001a3ba8?, 0x0?})
	/usr/lib/g
```

not sure if this one is connected to the first one, or what causes it.

Reply:
> I mean we would be returning a type which is not exported and does not give a clear contract to the user what to expect. So using an interface seems like a better option.

I'm still not sure about this, this thread https://www.reddit.com/r/golang/comments/ohexln/trying_to_understand_dont_return_interface/ throws up some of the arguments for returning a concrete type.

Reply:
Hi @atiratree

Thanks for the info on "stress".  I've run and found a race where it was possible for the clock to tick either to or past the time of the nextReady item.  In this scenario the nextReady channel will not fire.  I've been able to mitigate the race by placing a check for the nextReady item being ready directly before the select (and wait) and continue round the loop again, if there are ready items to pop.

In real life, outside of fake clocks and tests I would have expected this to have been rare and if it did happen then the heartbeat signal would have fired after 10 seconds, triggering the waitLoop to process the items again.

Reply:
seems I am still getting the 2nd flake (points to TestForceQueueShutdownUsingShutDown), but can only reproduce it when I run it on the whole package.

@davidmccormick can you reproduce it, or should I try to debug it?

Reply:
correction: can reproduce it with `stress ./workqueue.test -test.run TestForceQueueShutdownUsingShutDown` but only got  first failure after 10m

```
10m20s: 148958 runs so far, 1 failures
```